File: hummingbird/androidApp/app/src/main/java/net/acumenus/hummingbird/ui/components/KpiTile.kt
package net.acumenus.hummingbird.ui.components

import androidx.compose.foundation.background
import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.IntrinsicSize
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxHeight
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import net.acumenus.hummingbird.data.CensusUnit
import net.acumenus.hummingbird.ui.theme.Z

/** Per-unit census tile: status stripe + name + chip + occupied/safe metric + occupancy bar. */
@Composable
fun KpiTile(unit: CensusUnit) {
    val status = unit.capacity
    Row(
        modifier = Modifier
            .fillMaxWidth()
            .height(IntrinsicSize.Min)
            .panel(),
    ) {
        Box(Modifier.width(4.dp).fillMaxHeight().background(status.color))
        Column(
            Modifier.padding(16.dp),
            verticalArrangement = Arrangement.spacedBy(8.dp),
        ) {
            Row(verticalAlignment = Alignment.Top) {
                Column(Modifier.weight(1f)) {
                    Text(unit.name, color = Z.ink, fontSize = 16.sp, fontWeight = FontWeight.SemiBold)
                    Text(
                        unit.type.replace("_", " ").uppercase(),
                        color = Z.inkMuted, fontSize = 10.sp, fontWeight = FontWeight.Medium, letterSpacing = 0.5.sp,
                    )
                }
                StatusChip(status)
            }

            Row(verticalAlignment = Alignment.Bottom, horizontalArrangement = Arrangement.spacedBy(8.dp)) {
                Text("${unit.occupied}", color = Z.ink, fontSize = 34.sp, fontWeight = FontWeight.SemiBold)
                Text("/ ${unit.safeCapacity} safe beds", color = Z.inkMuted, fontSize = 13.sp, modifier = Modifier.padding(bottom = 5.dp))
                Spacer(Modifier.weight(1f))
                if (unit.bedNeed > 0) {
                    Text("${unit.bedNeed} over", color = Z.statusCritical, fontSize = 12.sp, fontWeight = FontWeight.SemiBold, modifier = Modifier.padding(bottom = 5.dp))
                }
            }

            val ratio = if (unit.safeCapacity > 0) (unit.occupied.toFloat() / unit.safeCapacity).coerceIn(0f, 1f) else 0f
            Box(
                Modifier.fillMaxWidth().height(6.dp).clip(RoundedCornerShape(50)).background(Z.border),
            ) {
                Box(Modifier.fillMaxWidth(ratio).height(6.dp).clip(RoundedCornerShape(50)).background(status.color))
            }

            Row(horizontalArrangement = Arrangement.spacedBy(16.dp)) {
                Metric("${unit.available}", "available")
                Metric("${unit.blocked}", "blocked/dirty")
            }
        }
    }
}

@Composable
private fun Metric(value: String, label: String) {
    Row(horizontalArrangement = Arrangement.spacedBy(4.dp)) {
        Text(value, color = Z.ink, fontSize = 14.sp, fontWeight = FontWeight.SemiBold)
        Text(label, color = Z.inkMuted, fontSize = 12.sp)
    }
}

File: hummingbird/iosApp/Hummingbird/DesignSystem/Components/KpiTile.swift
import SwiftUI

/// A per-unit census tile: a left status stripe, the unit name + status chip, the big
/// occupied/safe metric (tabular figures), and a supporting available/blocked/bed-need line.
/// Mirrors the web KpiTile signature (status stripe + arrow/label, metrics in tabular-nums).
struct KpiTile: View {
    let unit: CensusUnit

    private var status: CapacityStatus { unit.capacity }

    var body: some View {
        HStack(spacing: 0) {
            // 4pt status stripe
            Rectangle()
                .fill(Z.status(status))
                .frame(width: 4)

            VStack(alignment: .leading, spacing: Z.s2) {
                HStack(alignment: .top) {
                    VStack(alignment: .leading, spacing: 2) {
                        Text(unit.name)
                            .font(Z.scaledFont(16, weight: .semibold))
                            .foregroundStyle(Z.ink)
                        Text(unit.type.replacingOccurrences(of: "_", with: " ").uppercased())
                            .font(Z.scaledFont(10, weight: .medium))
                            .tracking(0.5)
                            .foregroundStyle(Z.inkMuted)
                    }
                    Spacer()
                    StatusChip(status: status)
                }

                HStack(alignment: .firstTextBaseline, spacing: Z.s2) {
                    Text("\(unit.occupied)")
                        .font(Z.scaledFont(34, weight: .semibold))
                        .monospacedDigit()
                        .foregroundStyle(Z.ink)
                    Text("/ \(unit.staffedBedCount) staffed beds")
                        .font(Z.scaledFont(13, weight: .regular)).monospacedDigit()
                        .foregroundStyle(Z.inkMuted)
                    Spacer()
                    // "Over" only when occupancy exceeds staffed beds (bed_need > 0).
                    if unit.bedNeed > 0 && unit.staffedBedCount > 0 {
                        Label("\(unit.bedNeed) over", systemImage: "arrow.up")
                            .font(Z.scaledFont(12, weight: .semibold))
                            .foregroundStyle(Z.status(.critical))
                            .labelStyle(.titleAndIcon)
                    }
                }

                // Occupancy bar
                GeometryReader { geo in
                    let ratio = unit.staffedBedCount > 0 ? min(1.2, Double(unit.occupied) / Double(unit.staffedBedCount)) : 0
                    ZStack(alignment: .leading) {
                        Capsule().fill(Z.border)
                        Capsule()
                            .fill(Z.status(status))
                            .frame(width: max(2, geo.size.width * min(1.0, ratio)))
                    }
                }
                .frame(height: 6)

                HStack(spacing: Z.s3) {
                    metric("\(unit.available)", "available")
                    metric("\(unit.blocked)", "blocked/dirty")
                    metric("\(unit.canAdmit)", "can admit")
                    Spacer()
                }
            }
            .padding(Z.s4)
        }
        .background(RoundedRectangle(cornerRadius: Z.radius, style: .continuous).fill(Z.surface))
        .overlay(RoundedRectangle(cornerRadius: Z.radius, style: .continuous).strokeBorder(Z.border, lineWidth: 1))
        .clipShape(RoundedRectangle(cornerRadius: Z.radius, style: .continuous))
        .shadow(color: .black.opacity(0.25), radius: 8, x: 0, y: 2)
        .accessibilityElement(children: .combine)
        .accessibilityLabel("\(unit.name), \(status.label), \(unit.occupied) of \(unit.staffedBedCount) staffed beds occupied, can admit \(unit.canAdmit)")
    }

    private func metric(_ value: String, _ label: String) -> some View {
        HStack(spacing: Z.s1) {
            Text(value).font(Z.scaledFont(14, weight: .semibold)).monospacedDigit().foregroundStyle(Z.ink)
            Text(label).font(Z.scaledFont(12)).foregroundStyle(Z.inkMuted)
        }
    }
}
